src/passes/Unsubtyping.cpp

```diff
@@ -0,0 +1,580 @@
+        auto it = worklist.begin();
+        auto type = *it;
+        worklist.erase(it);
+        auto super = supertypes.at(type);
+        if (super.isBasic()) {
+          continue;
+        }
+
+        if (type.isStruct()) {
+          const auto& fields = type.getStruct().fields;
+          const auto& superFields = super.getStruct().fields;
+          for (size_t i = 0, size = superFields.size(); i < size; ++i) {
+            if (fields[i].mutable_ == Immutable) {
+              noteSubtype(fields[i].type, superFields[i].type);
+            }
+          }
+        } else if (type.isArray()) {
+          auto elem = type.getArray().element;
+          if (elem.mutable_ == Immutable) {
+            noteSubtype(elem.type, super.getArray().element.type);
+          }
+        } else {
+          assert(type.isSignature());
+          auto sig = type.getSignature();
+          auto superSig = super.getSignature();
+          noteSubtype(superSig.params, sig.params);
+          noteSubtype(sig.results, superSig.results);
+        }
+      }
+
+      // Analyze all casts at once.
+      // TODO: This is expensive. Analyze casts incrementally after we
+      // initially analyze them.
+      analyzeCasts();
+    }
+  }
+
+  void analyzeCasts() {
+    // For each cast (src, dest) pair, any type that remains a subtype of src
+    // (meaning its values can inhabit locations typed src) and that was
+    // originally a subtype of dest (meaning its values would have passed the
+    // cast) should remain a subtype of dest so that its values continue to pass
+    // the cast.
+    //
+    // For every type, walk up its new supertype chain to find cast sources and
+    // compare against their associated cast destinations.
+    for (auto it = supertypes.begin(); it != supertypes.end(); ++it) {
+      auto type = it->first;
+      for (auto superIt = it; superIt != supertypes.end();
+           superIt = supertypes.find(superIt->second)) {
+        auto super = superIt->second;
+        auto destsIt = castTypes.find(super);
+        if (destsIt == castTypes.end()) {
+          continue;
+        }
+        for (auto dest : destsIt->second) {
+          if (HeapType::isSubType(type, dest)) {
+            noteSubtype(type, dest);
+          }
+        }
+      }
+    }
+  }
+
+  void optimizeTypes(Module& wasm) {
+    struct Rewriter : GlobalTypeRewriter {
+      Unsubtyping& parent;
+      Rewriter(Unsubtyping& parent, Module& wasm)
+        : GlobalTypeRewriter(wasm), parent(parent) {}
+      std::optional<HeapType> getSuperType(HeapType type) override {
+        if (auto it = parent.supertypes.find(type);
+            it != parent.supertypes.end() && !it->second.isBasic()) {
+          return it->second;
+        }
+        return std::nullopt;
+      }
+    };
+    Rewriter(*this, wasm).update();
+  }
+
+  void doWalkModule(Module* wasm) {
+    // Visit the functions in parallel.
+    ModuleUtils::ParallelFunctionAnalysis<Unsubtyping> analysis(
```

Comment:
It's a little awkward to create an entire Unsubtyping for each function, including the `worklist`. Though maybe that's the only field that isn't needed in both parallel and later work, so maybe it's ok?

Reply:
Yes, and importantly it lets us reuse all the `noteSubtype` and `noteCast` helper functions rather than having to duplicate them.

Reply:
ohhh, that's a good idea. I was struggling to come up with a good solution there. Still, I'm not convinced that anything less than the whole set of subtypes we calculate here can be useful.